website/blog/2021-08-18-improving-marker-mechanism.md

```diff
@@ -0,0 +1,65 @@
+
+As the number of data files to write increases, so does the number of marker files to create.  This can create performance bottlenecks for cloud storage such as AWS S3.  In AWS S3, each file create and delete call triggers an HTTP request and there is [rate-limiting](https://docs.aws.amazon.com/AmazonS3/latest/userguide/optimizing-performance.html) on how many requests can be processed per second per prefix in a bucket.  When the number of data files to write concurrently and the number of marker files is huge, the marker file operations become the performance bottleneck.  In one case, the marker file deletion takes an hour to finish due to S3 rate-limiting for a bulk insert operation running for a few hours.  Such behavior degrades the performance of the write.
+
+## Timeline-server-based marker mechanism improving write performance
+
+To address the performance bottleneck due to rate-limiting of AWS S3 explained above, we introduce a **new marker mechanism leveraging the timeline server**, which optimizes the marker-related latency for file systems with non-trivial file I/O latency.  The **timeline server** in Hudi serves as a centralized place for providing the file system and timeline views. As shown below, the new timeline-server-based marker mechanism delegates the marker creation and other marker-related operations from individual executors to the timeline server for centralized processing.  The timeline server maintains the created markers in memory for corresponding marker requests.  The timeline server achieves consistency by periodically flushing the in-memory markers to a limited number of underlying files in the file system.  In such a way, the number of actual file operations and latency related to markers can be significantly reduced even with a huge number of data files, thus improving the performance of the writes.
+
+![Timeline-server-based marker mechanism](/assets/images/blog/marker-mechanism/timeline-server-based-marker-mechanism.png)
+
+To improve the efficiency of processing marker creation requests, we design the batch processing in the handler of marker requests at the timeline server.  Each marker creation request is handled asynchronously in the Javalin timeline server and queued before processing. For every batch interval, e.g., 20ms, a dispatching thread pulls the pending requests from the queue and sends them to the worker thread for processing. Each worker thread processes the marker creation requests, sets the responses, and flushes the new markers by overwriting the underlying file storing the markers in the file system.  There are multiple worker threads running concurrently, given that the file overwriting takes longer than the batch interval, and each worker thread writes to an exclusive file not touched by other threads, to guarantee consistency and correctness. Both the batch interval and the number of worker threads can be configured through the write options.
```

Comment:
I don't think we need to go into worker threads as such(sets responses, etc). We can say that timeline server marker handler queues marker requests every 20 ms and writes all markers to next marker file in a round robin fashion. don't think we need to get into impl details.